### ISSUE TITLE
Translatable components

### ISSUE THREAD

**Issue by fantoine** [supertokens/supertokens-auth-react#233]
Hi here,

I was wondering if there is a way to translate the components ?

One way could be to let us override some components to change the text, or just give us access to some keys that we could override.

**Comment by rishabhpoddar** [comment]
Great idea! Will figure out the best way to enable this feature. 

Until then, you can modify the UI by overriding it the following way:

- Disabling the default UI for the sign in and all other pages by setting `disableDefaultImplementation` to true in `signInAndUpFeature` (and other features) when calling the `init` function (see https://supertokens.io/docs/auth-react/emailpassword/init). This will basically remove the UI from `/auth/*`. This mean you can handle this route to render anything you like.

- Then you would need to create your own component to show the login / reset pass etc UIs. For example, the login screen:

```js
import {SignInAndUp} from "supertokens-auth-react/recipe/emailpassword";

render() {
  return (
    <SignInAndUp>
      <YOUR UI COMPONENT GOES HERE>
    </SignInAndUp>
  );
}
```
- The default UI component can be found [in this source file](https://github.com/supertokens/supertokens-auth-react/blob/master/lib/ts/recipe/emailpassword/components/themes/signInAndUp/index.tsx) (for emailpassword recipe). So please feel free to copy this and modify the text to be in the language you like.

**Comment by rishabhpoddar** [comment]
How do you imagine we could add this feature of translations? Any ideas please?

**Comment by fantoine** [comment]
Thanks for the tip to translate components !

For translations, I guess you could add some `translation` object  to each recipes which need some texts.
It's also possible to integrate directly some libs like react-i18next but it may not be suitable for all projects.

What do you think about it ?

**Comment by fantoine** [comment]
Have you also explored the possibility to use a React context instead of the `SuperTokens.init` singleton to share informations between components ?

**Comment by rishabhpoddar** [comment]
Thanks for recommending `react-i18next`. I will check it out.

We wanted to use `SuperTokens.init` singleton because:
- The docs for it would be similar to that of initing supertokens on the backend (nodejs), or any other non-react framework (like angular or vue). 
- A recipe is defined as "more than a component". For example, the session recipe adds interceptors to `axios` or `fetch` and that should ideally be added independently of loading any component in the app.

Are you facing any issues due to the way we have designed this aspect of the lib?

**Comment by fantoine** [comment]
No, I don't have any specific issue right now with this.
I get your point of view and agree with you about keeping a cross-framework way to do this. :)

**Comment by rishabhpoddar** [comment]
Until this is natively supported, a way around getting translatable components is to create your own theme from our theme (Assume that you are using `emailpassword` recipe):

## Step 1:
Disable all the default UI pages we provide by setting disableDefaultImplementation to true on the frontend config like so:
```js
SuperTokens.init({
    appInfo: {...},
    recipeList: [
        EmailPassword.init({
            signInAndUpFeature: {
                disableDefaultImplementation: false,
                ...
            },
            resetPasswordUsingTokenFeature: {
                disableDefaultImplementation: false
                ...
            },
            emailVerificationFeature: {
                disableDefaultImplementation: false
                ...
            },
        })
    ]
});
```

After you do the above, going to `/auth` should not show the login UI.

## Step 2:
Create a login page on `/auth`, and in the render function, do the following:

```js
import {SignInAndUp} from "supertokens-auth-react/recipe/emailpassword"

render() {
  <SignInAndUp> // this is the "sign in feature component"
    <YourLoginThemeHere />
  </SignInAndUp>
}
```

To make `YourLoginThemeHere`, copy the contents of [this folder](https://github.com/supertokens/supertokens-auth-react/tree/master/lib/ts/recipe/emailpassword/components/themes) into your own project, change TS to JS if you need to, and fix the import statements in all the files to point to our lib.

Then `YourLoginThemeHere` component essentially points to the component in [this file](https://github.com/supertokens/supertokens-auth-react/blob/master/lib/ts/recipe/emailpassword/components/themes/signInAndUp/index.tsx).

You can then change the text in that file and other files to be in a different language.

## Step 3:
Do something similar for reset password screen. The reset password path is `/auth/reset` and the feature component for that is `import {ResetPasswordUsingToken} from "supertokens-auth-react/recipe/emailpassword"`. The theme component for that is in [this file](https://github.com/supertokens/supertokens-auth-react/blob/master/lib/ts/recipe/emailpassword/components/themes/resetPasswordUsingToken/index.tsx)

## Step 4:
Do something similar for email verification screen. The reset password path is `/verify-email` and the feature component for that is here `import {EmailVerification} from "supertokens-auth-react/recipe/emailpassword"`. The theme component for that is [this file](https://github.com/supertokens/supertokens-auth-react/blob/master/lib/ts/recipe/emailverification/components/themes/emailVerification/index.tsx)

**Comment by rishabhpoddar** [comment]
IMPORTANT: We also need to keep in mind that language choice can be dynamically loaded which implies that it may be fetched post the `init` call.

**Comment by rishabhpoddar** [comment]
Things to consider:
- [ ] We want to fit into how people do translations.
- [ ] Change language dynamically.
   - Enable adding extra languages dynamically
- [ ] Easily allow a user to know which strings they need to provide for their language and view all the default copy in one place.
    - Translation keys should be constructed in a way that points directly to where they are used.
- [ ] Being able to use other translations lib along with us.
- [ ] Sending translation keys from APIs instead of actual error messages. This would be in conjunction with current API output. So for example, right now we have (in thirdparty recipe) `{status: "FIELD_ERROR", message: string}`. So this would change to `{status: "FIELD_ERROR", message: string, translationKey: string}`.
- [ ] Dynamic message construction?
- [ ] For form fields labels we first assume that they are translation keys, and if they aren't we just show them as is.
- [ ] Try and avoid using react-i18next or any other library.

**Comment by rishabhpoddar** [comment]
This feature has been released in version 0.18.6.
